onnxruntime/core/optimizer/nchwc_transformer.cc

```diff
@@ -0,0 +1,721 @@
+    // is decremented as uses are converted to NCHWc format. Nodes are inserted
+    // to reorder the output if this count is non-zero.
+    size_t remaining_original_uses_;
+
+    // Stores the logical number of channels for this NCHWc output. The NCHWc
+    // NodeArg is zero padded to the NCHWc block size. If the output needs to
+    // be reordered back to a standard tensor format, this channel count is
+    // used to generate the expected number of channels.
+    const int64_t channels_;
+
+    // Stores the proto shape for the NCHWc output.
+    NchwcArgument::Shape shape_;
+
+    NchwcArgument(Node& output_node, NodeArg* output_nchwc_arg, size_t original_uses, size_t channels, const NchwcArgument::Shape& shape)
+        : output_node_(output_node),
+          nchwc_arg_(output_nchwc_arg),
+          starting_original_uses_(original_uses),
+          remaining_original_uses_(original_uses),
+          channels_(channels),
+          shape_(shape) {
+    }
+  };
+
+  const ONNX_NAMESPACE::AttributeProto* GetAttribute(const Node& node, const char* attribute_name);
+  size_t RemoveOutputEdges(Node& node);
+  void CreateNchwcArgument(Node& node, Node& nchwc_node, int64_t channels, const NchwcArgument::Shape& shape);
+  void FuseNchwcArgument(Node& node, const NchwcArgument& nchwc_arg);
+  void InsertReorderInput(Node& node);
+
+  void ConvPoolShapeInference(const Node& node,
+                              const NchwcArgument::Shape& input_shape,
+                              NchwcArgument::Shape& output_shape,
+                              const ONNX_NAMESPACE::TensorProto* filter_shape);
+
+  void TransformConv(Node& node);
+  void TransformPool(Node& node);
+  void TransformAdd(Node& node);
+  void TransformConcat(Node& node);
+  void TransformActivation(Node& node);
+  void TransformElementwise(Node& node);
+
+  Graph& graph_;
+
+  // Stores a queue of nodes to be removed after walking through the graph.
+  std::deque<NodeIndex> removed_nodes_;
+
+  // Stores a mapping from the original NodeArg outputs to the NCHWc variants
+  // created inside this graph transform.
+  std::unordered_map<NodeArg*, std::unique_ptr<NchwcArgument>> nchwc_args_;
+
+  // Stores a mapping of NodeArg inputs that have already been reordered, so
+  // multiple nodes can share the NCHWc input.
+  std::unordered_map<NodeArg*, NodeArg*> reorder_inputs_;
+
+  // Stores a mapping of NodeArg filters that have already been reordered, so
+  // multiple nodes can share the NCHWc filter.
+  std::unordered_map<NodeArg*, NodeArg*> filters_OIHWBo_;
+  std::unordered_map<NodeArg*, NodeArg*> filters_OIHWBiBo_;
+
+  // Stores a mapping of NodeArg biases that have already been aligned to the
+  // NCHWc block size, so multiple nodes can share the NCHWc biases.
+  std::unordered_map<NodeArg*, NodeArg*> aligned_biases_;
+};
+
+const ONNX_NAMESPACE::AttributeProto* NchwcTransformerImpl::GetAttribute(const Node& node, const char* attribute_name) {
```

Comment:
This seems pretty generic. Should Node have this method, or could it be in graph_utils?

Reply:
Maybe something for the optimizer team to do if they like, but I would prefer not to deal with this in this PR.

Reply:
I think we can use the existing graph_utils::GetNodeAttribute() here?